File: agent_mcp/core/globals.py
# Agent-MCP/mcp_template/mcp_server_src/core/globals.py
"""
Centralized mutable global state for the MCP server.
To use:
from mcp_server_src.core import globals as g
g.admin_token = "new_token"
"""
import anyio # For rag_index_task type hint
from typing import Dict, List, Optional, Any

# --- Core Server State ---
# From main.py:147
# Client ID -> Connection data (Note: original usage of 'connections' might be simplified
# or its management moved if it's purely for SSE connection tracking by the transport layer)
connections: Dict[str, Any] = {}

# From main.py:148
active_agents: Dict[str, Dict[str, Any]] = {}  # Agent Token -> Agent data

# From main.py:149
# This is the runtime admin_token.
# Initialization logic (generate/load) will be handled during server startup.
admin_token: Optional[str] = None

# From main.py:150
tasks: Dict[str, Dict[str, Any]] = {}  # Task ID -> Task data (in-memory cache of tasks)

# --- File and Directory State ---
# From main.py:153
file_map: Dict[str, Dict[str, Any]] = {}  # filepath -> {"agent_id": ..., "timestamp": ..., "status": ...}

# From main.py:154
agent_working_dirs: Dict[str, str] = {}  # agent_id -> absolute_working_directory_path

# --- Auditing and Agent Management ---
# From main.py:155
# In-memory audit log for the current session. Persistent log is 'agent_audit.log'.
audit_log: List[Dict[str, Any]] = []

# From main.py:158
agent_profile_counter: int = 20 # For cycling Cursor profile numbers

# From main.py:166
agent_color_index: int = 0 # For cycling through AGENT_COLORS from config.py

# --- Server Lifecycle ---
# From main.py:169
server_running: bool = True # Flag to control main server loop and background tasks, handled by signal_utils.py

# Flag to track if server initialization is complete
# This prevents handling requests before critical components are ready
server_initialized: bool = False

# Timestamp when server initialization completed
server_start_time: Optional[float] = None

# --- External Service Clients (Placeholders) ---
# From main.py:185
# The actual OpenAI client instance will be initialized and managed by external/openai_service.py.
# This global variable can serve as a reference if truly global access is needed,
# though passing the client explicitly or using a getter from openai_service is cleaner.
# For now, we'll keep it as a placeholder reflecting the original structure.
# Type hint can be refined to `openai.OpenAI` once that module is structured.
openai_client_instance: Optional[Any] = None

# Async OpenAI client instance for use in async contexts
# This avoids blocking the event loop when making OpenAI API calls
openai_async_client_instance: Optional[Any] = None

# --- Database/VSS State ---
# From main.py:200
# Flag to check if sqlite-vec extension loadability has been tested.
global_vss_load_tested: bool = False

# From main.py:201
# Flag indicating if sqlite-vec extension was successfully loaded during the initial test.
global_vss_load_successful: bool = False

# --- Background Task Handles ---
# From main.py:510 (and used in main.py:1943, 2627, 2641)
# Handle for the RAG indexing background task, typically managed by an anyio.TaskGroup.
# The type hint `anyio.abc.CancelScope` is a common way to hold a reference that allows cancellation.
rag_index_task_scope: Optional[anyio.abc.CancelScope] = None

# Note: The original `main.py` also had `openai_client = None` at line 185.
# I've named it `openai_client_instance` here to avoid confusion with the module name
# if we later have `import openai_client from ...`.
# The actual OpenAI client will be initialized and managed in `external/openai_service.py`.

File: agent-mcp-node/src/core/globals.ts
// Global state management for Agent-MCP Node.js
// Ported from Python core/globals.py

/**
 * Centralized mutable global state for the MCP server
 * This replaces the Python globals module with TypeScript equivalents
 */

// Agent data interface
interface AgentData {
  agent_id: string;
  capabilities: string[];
  status: 'created' | 'active' | 'terminated' | 'failed';
  current_task?: string;
  working_directory: string;
  color: string;
  created_at: string;
  updated_at: string;
}

// File lock info interface
interface FileLockInfo {
  agent_id: string;
  timestamp: string;
  status: 'locked' | 'unlocked';
  operation?: string;
}

// Task data interface (for in-memory cache)
interface TaskData {
  task_id: string;
  title: string;
  description: string;
  assigned_to?: string;
  created_by: string;
  status: string;
  priority: string;
  created_at: string;
  updated_at: string;
  parent_task?: string;
  child_tasks: string[];
  depends_on_tasks: string[];
  notes: any[];
}

// Audit log entry interface
interface AuditLogEntry {
  timestamp: string;
  agent_id: string;
  action: string;
  details: any;
}

/**
 * Global state container
 * Maintains all server-wide state that needs to persist across requests
 */
export const globalState = {
  // --- Core Server State ---
  
  /**
   * Client connections tracking
   * Maps client ID to connection data
   */
  connections: new Map<string, any>(),
  
  /**
   * Active agents registry
   * Maps agent token to agent data
   */
  activeAgents: new Map<string, AgentData>(),
  
  /**
   * Runtime admin token
   * Generated during server startup or loaded from environment
   */
  adminToken: null as string | null,
  
  /**
   * Tasks in-memory cache
   * Maps task ID to task data for quick access
   */
  tasks: new Map<string, TaskData>(),
  
  // --- File and Directory State ---
  
  /**
   * File locking state
   * Maps file path to lock information
   */
  fileMap: new Map<string, FileLockInfo>(),
  
  /**
   * Agent working directories
   * Maps agent ID to absolute working directory path
   */
  agentWorkingDirs: new Map<string, string>(),
  
  // --- Tmux Session Management ---
  
  /**
   * Active tmux sessions
   * Maps agent ID to tmux session name
   */
  agentTmuxSessions: new Map<string, string>(),
  
  // --- Auditing and Agent Management ---
  
  /**
   * In-memory audit log for current session
   * Persistent log should be stored in files/database
   */
  auditLog: [] as AuditLogEntry[],
  
  /**
   * Agent profile counter for cycling Cursor profiles
   */
  agentProfileCounter: 20,
  
  /**
   * Agent color index for cycling through available colors
   */
  agentColorIndex: 0,
  
  // --- Server Lifecycle ---
  
  /**
   * Server running flag
   * Controls main server loop and background tasks
   */
  serverRunning: true,
  
  /**
   * Server start time for uptime calculations
   */
  serverStartTime: new Date().toISOString(),
  
  // --- External Service Clients ---
  
  /**
   * OpenAI client instance placeholder
   * Will be initialized by external service modules
   */
  openaiClientInstance: null as any,
  
  // --- Database/VSS State ---
  
  /**
   * VSS load test status
   * Tracks if sqlite-vec extension loadability has been tested
   */
  vssLoadTested: false,
  
  /**
   * VSS load success status
   * Indicates if sqlite-vec extension was successfully loaded
   */
  vssLoadSuccessful: false,
  
  // --- Background Task Handles ---
  
  /**
   * RAG indexing background task handle
   * Used for cancellation and lifecycle management
   */
  ragIndexTaskHandle: null as any,
  
  /**
   * Claude Code session monitoring task handle
   */
  claudeSessionTaskHandle: null as any,
  
  // --- Statistics and Monitoring ---
  
  /**
   * Request statistics
   */
  stats: {
    totalRequests: 0,
    totalToolCalls: 0,
    totalErrors: 0,
    startTime: Date.now()
  }
};

/**
 * Helper functions for global state management
 */

/**
 * Reset global state (useful for testing)
 */
export function resetGlobalState(): void {
  globalState.connections.clear();
  globalState.activeAgents.clear();
  globalState.tasks.clear();
  globalState.fileMap.clear();
  globalState.agentWorkingDirs.clear();
  globalState.agentTmuxSessions.clear();
  globalState.auditLog.length = 0;
  globalState.agentProfileCounter = 20;
  globalState.agentColorIndex = 0;
  globalState.serverRunning = true;
  globalState.stats.totalRequests = 0;
  globalState.stats.totalToolCalls = 0;
  globalState.stats.totalErrors = 0;
  globalState.stats.startTime = Date.now();
}

/**
 * Get server statistics
 */
export function getServerStats() {
  const uptime = Date.now() - globalState.stats.startTime;
  
  return {
    uptime,
    uptimeHours: Math.floor(uptime / (1000 * 60 * 60)),
    uptimeMinutes: Math.floor((uptime % (1000 * 60 * 60)) / (1000 * 60)),
    totalRequests: globalState.stats.totalRequests,
    totalToolCalls: globalState.stats.totalToolCalls,
    totalErrors: globalState.stats.totalErrors,
    activeAgents: globalState.activeAgents.size,
    activeTasks: globalState.tasks.size,
    lockedFiles: globalState.fileMap.size,
    tmuxSessions: globalState.agentTmuxSessions.size,
    auditEntries: globalState.auditLog.length
  };
}

/**
 * Increment request counter
 */
export function incrementRequestCount(): void {
  globalState.stats.totalRequests++;
}

/**
 * Increment tool call counter
 */
export function incrementToolCallCount(): void {
  globalState.stats.totalToolCalls++;
}

/**
 * Increment error counter
 */
export function incrementErrorCount(): void {
  globalState.stats.totalErrors++;
}

/**
 * Add entry to audit log
 */
export function addAuditLogEntry(agentId: string, action: string, details: any = {}): void {
  globalState.auditLog.push({
    timestamp: new Date().toISOString(),
    agent_id: agentId,
    action,
    details
  });
  
  // Keep audit log size reasonable (last 1000 entries)
  if (globalState.auditLog.length > 1000) {
    globalState.auditLog.splice(0, globalState.auditLog.length - 1000);
  }
}

/**
 * Get recent audit log entries
 */
export function getAuditLog(limit: number = 50, agentId?: string, action?: string): AuditLogEntry[] {
  let filtered = globalState.auditLog;
  
  if (agentId) {
    filtered = filtered.filter(entry => entry.agent_id === agentId);
  }
  
  if (action) {
    filtered = filtered.filter(entry => entry.action === action);
  }
  
  return filtered.slice(-limit);
}

/**
 * Update task cache
 */
export function updateTaskCache(taskId: string, taskData: TaskData): void {
  globalState.tasks.set(taskId, taskData);
}

/**
 * Remove task from cache
 */
export function removeTaskFromCache(taskId: string): void {
  globalState.tasks.delete(taskId);
}

/**
 * Get task from cache
 */
export function getTaskFromCache(taskId: string): TaskData | undefined {
  return globalState.tasks.get(taskId);
}

console.log('✅ Global state management loaded');